8353440: Disable FTP fallback for non-local file URLs by default

eirbjo opened the pull request:
Please help review this PR which disables the unspecified but long-standing feature where an `FtpURLConnection` is opened as a fallback for non-local file URLs.

Before this change, if a file URL has a non-local host component, say `file://remotehost/folder/data.txt`, then the  implementation would attempt opening an FTP connection to `remotehost`. After this change, such URLs will be rejected with a `MalformedURLException`, unless the FTP fallback feature is explicitly re-enabled via a system property.

This change was initially discussed here: https://mail.openjdk.org/pipermail/net-dev/2025-March/025988.html

See the above discussion and CSR draft JDK-8354678 for the motivation for this change.

This PR:

* Changes file URL `Handler::openConnection` implementation for unix/windows to throw `MalformedURLException`, unless the FTP fallback feature is explicitly enabled by configuration.
* Introduces a new system property `jdk.net.file.ftpenabled` which when set to `true` re-enables the feature.
* Documents the new property in `net-properties.html` 
* Updates the existing test `NonLocalFtpFallback` to enable the feature via said system property.
* Adds a new test `NonLocalFtpFallbackDisabled` verifying that a `MalformedURLException` is thrown by default for a non-local URL host component.
* Moves testing of exceptional behavior in FtpConnection when using non-local file URLs with FTP fallback enabled from `OpenStream` to `NonLocalFtpFallback`

I have added a Release Note as a subtask in the JBS issue, this also needs a review.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8354678](https://bugs.openjdk.org/browse/JDK-8354678) to be approved

### Issues
 * [JDK-8353440](https://bugs.openjdk.org/browse/JDK-8353440): Disable FTP fallback for non-local file URLs by default (**Enhancement** - P4)
 * [JDK-8354678](https://bugs.openjdk.org/browse/JDK-8354678): Disable FTP fallback for non-local file URLs by default (**CSR**)


### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24657/head:pull/24657` \
`$ git checkout pull/24657`

Update a local copy of the PR: \
`$ git checkout pull/24657` \
`$ git pull https://git.openjdk.org/jdk.git pull/24657/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24657`

View PR using the GUI difftool: \
`$ git pr show -t 24657`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24657.diff">https://git.openjdk.org/jdk/pull/24657.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24657#issuecomment-2806339314)
</details>

eirbjo commented:
/csr

ecki commented:
Is there a Risk that This breaks Applications which parses Potential UNC path as URL (or even worse want to open them?. I am not sure if this actually ever worked, but) I know people are reguläres confused about file:/// (when is it needed), and even Chat gpt suggests:

> UNC path → \\server\share\file.txt
> File URL → file://server/share/file.txt

eirbjo commented:
> Is there a Risk that This breaks Applications which parses Potential UNC path as URL (or even worse want to open them?. I am not sure if this actually ever worked, but) I know people are reguläres confused about file:/// (when is it needed), 

Thanks for rassing this important concern!

The current behavior for _existing_ UNC paths would be unaffected by this path. The testing for UNC paths (by converting the path to UNC form and using `File::exists`) happens before the FTP fallback. See the Windows implementation of Handler:openConnection.

I've added a mention of this in the release notes.

eirbjo commented:
> The current behavior for _existing_ UNC paths would be unaffected by this change.

For the sake of completeness: 

If `Files::exists` failed for a UNC path before this change, then the code would continue to the FTP fallback and the result would depend on whether a public anonymous FTP server was available on the non-local host. 

After this change, when a UNC path does not exist, then the connection will fail with a `MalformedURLException`, just like the Unix implementation.

eirbjo commented on src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java:
```diff
@@ -46,6 +47,12 @@ public class FileURLConnection extends URLConnection {
     private static final String TEXT_PLAIN = "text/plain";
     private static final String LAST_MODIFIED = "last-modified";
 
+    // The feature of falling back to FTP for non-local file URLs is disabled
+    // by default and can be re-enabled by setting a system property
+    private static boolean FTP_FALLBACK_ENABLED = "true".equalsIgnoreCase(
+            System.getProperty("sun.net.www.protocol.file.ftp-enabled", "false")
+    );
+
```
The code for defining and checking this system property would ideally belong in `Handler.java`.

I decided to put it in `FileURLConnection.java` instead, just to avoid increasing the already significant duplication of code between unix/windows implementations of `Handler.java`.

AlanBateman commented:
No objection to disabling this legacy feature by default but I think it will require writing a release note, and thus will document the system property. Right now, the proposal seems to be name the system property "sun.net.www.protocol.file.ftp-enabled" and I think we'll have to rename this to something starting with "jdk.net", and some combination of works that makes it clear that it allows file URL to make ftp connections. I can suggest some property names if you wish.

eirbjo commented:
> No objection to disabling this legacy feature by default but I think it will require writing a release note, and thus will document the system property

Indeed.  A provisional release note can be found here: https://bugs.openjdk.org/browse/JDK-8354658

> Right now, the proposal seems to be name the system property "sun.net.www.protocol.file.ftp-enabled" and I think we'll have to rename this to something starting with "jdk.net", and some combination of works that makes it clear that it allows file URL to make ftp connections.

Yes, in fact I did not put much thought into the name, since I anticipated reviewers to have input on this choice anyhow. 

> I can suggest some property names if you wish.

That would be great! I'm not sure what would be a good pattern/precedence to follow here. Anything you can propose would most likely be an improvement over the current "working title" :-)

Michael-Mc-Mahon commented:
No objection here to this change either. I too was wondering about how this might affect UNC behavior. But, I see that processing occurs before reaching this code. 

The property will need to be documented in src/java.base/share/classes/java/net/doc-files/net-properties.html
 also.

As for the name: how about something like "jdk.net.file.ftpfallback" ?

dfuch commented:
> I have made a first attempt of a text documenting the new property. Since I don't want the documentation of this feature to encourage wider use, I added some verbiage recommending that explicit `ftp://` URIs should be used instead.

Hi Eirik! 

Sorry for being late to the party. 
I too would welcome the removal of this old legacy feature. I may have more comments later (I haven't looked at the proposed code changes yet) - but here is a first feedback on the wording:

There are no handlers for URIs - only for URLs - so I would replace URI with URL in the property description to avoid any confusion. 

best regards,

-- daniel

eirbjo commented:
> Sorry for being late to the party. I too would welcome the removal of this old legacy feature.

We have not really discussed a future removal of this feature in the context of this PR. This PR simply disables the feature by default.

If we want to remove it in the future, should we include text now to mention it may be removed in a future release? Or is that something to consider for a future release?

dfuch commented:
> We have not really discussed a future removal of this feature in the context of this PR. This PR simply disables the feature by default.

Yes - sorry my bad. The disabling of this feature (and hopefully its removal in some future version of the JDK)

>If we want to remove it in the future, should we include text now to mention it may be removed in a future release? Or is that something to consider for a future release?

Good question. I believe disabling is a good first step, which will hopefully tell us whether anyone is actually using it.
We can decide later if/when we actually want to remove it.

dfuch commented:
> The Proxy section curiusly seems to use **URI** everywhere, but I've changed to using **URL** for this new section. It's a bit weird now that RFC 8089 uses URI, but maybe it's too hard to be consistent **and** correct in this case?

This is because `ProxySelector` thankfully uses URI (as in java.net.URI) not URL. So it's actually consistent *and* correct :-)

dfuch reviewed (commented):
This looks quite good. I will run it through our CI to see if anything falls from the tree.

eirbjo commented on src/java.base/share/classes/java/net/doc-files/net-properties.html:
```diff
@@ -169,6 +169,17 @@ <H2>Proxies</H2>
 </UL>
+<a id="FileHandler"></a>
+<H2>File URL stream protocol handler properties</H2>
+<P>The following properties are used to configure the handler for URLs with the {@code file://} scheme:</P>
```
> I wonder if it should say the `{@code file:}` scheme instead. 

This seems like an improvement to me. Not really a deliberate choice to include the slashes on my part, probably more for consistency with other schemes under Proxy. Files with any host component should be a rarity.

> Don't change anything right now - let's see if we have other opinions.

Sure, we can let this simmer a bit.

eirbjo replied:
> I wonder if it should say the `{@code file:}` scheme instead. Although I understand that the case that interests us here is "file://some.host/some/path". Don't change anything right now - let's see if we have other opinions.

Nobody commented on this. Should we keep the slashes - or slash them?

dfuch commented:
> I'm curious what the feelback channel would be here? Do vendors collect metrics for use of system properties, directly or indirectly through support channels such that usage of a property like this may be gauged somehow?

Nothing of the sort. But if something stops working we usually get a bug report. It's possible that application maintainers will discover the property, use it, and that we won't hear about it of course.

eirbjo commented on src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java:
```diff
@@ -222,4 +229,17 @@ public Permission getPermission() throws IOException {
+     */
+     static void requireFtpFallbackEnabled() throws MalformedURLException {
+        if (!FTP_FALLBACK_ENABLED) {
+            throw new MalformedURLException("Unsupported non-local file URL");
```
I aimed to keep this error message terse but informative, without revealing any parts of the URL spec.

I user stumbling on a rejected URL will not have a lot to work from here though..

If anyone have suggestions to improve the message, please chime in!

dfuch replied:
I have some ideas, but you may need to wait until https://git.openjdk.org/jdk/pull/23929 is integrated.

dfuch commented:
@eirbjo something must be missing on windows. Many tests under `java/net/URLConnection` are now failing on windows with the message: Unsupported non-local file URL

Here is the stack trace for UNCTest.java for instance:

```
java.net.MalformedURLException: Unsupported non-local file URL
	at java.base/sun.net.www.protocol.file.FileURLConnection.requireFtpFallbackEnabled(FileURLConnection.java:242)
	at java.base/sun.net.www.protocol.file.Handler.openConnection(Handler.java:91)
	at java.base/sun.net.www.protocol.file.Handler.openConnection(Handler.java:61)
	at java.base/java.net.URL.openConnection(URL.java:1206)
	at UNCTest.main(UNCTest.java:38)
```

eirbjo commented:
> java/net/URLConnection/UNCTest.java

It's not completely clear to me whether this test was originally written under the assumption that a FileURLConnection for an UNC path would be returned by this test.

If so, the CI instance running the test would need access to an existing UNC path like `\\jdk\LOCAL-JAVA/\dk1.4\win\README.txt`.

My wild guess would be that any current CI system doesn't have this particular UNC share/file available, so instead an (unconnected) FtpURLConnection would be returned. If that's the case, then I don't think this test actually excercises the code path intended. That would happen if the UNC path exists and a FileURLConnection was returned.

To summarize: This test wants to verify that any URLConnection to an UNC path is unconnected. Instead it just tests that the FtpURLConnection is unconnected, which it always is and has been and which in any case seems unrelated to JDK-4401485 

Or am I missing something here?

eirbjo commented:
> > java/net/URLConnection/UNCTest.java

> My wild guess would be that any current CI system doesn't have this particular UNC share/file available, so instead an (unconnected) FtpURLConnection would be returned. If that's the case, then I don't think this test actually excercises the code path intended. That could only happen if the UNC path exists and a FileURLConnection was returned.

As a side note:

If Handler::openConnection was updated to use `java.nio.file.Files::exists` instead of `java.io.File:exists`, then we could rewrite such tests that now rely on the UNC environment to instead mock the UNC path into existance using a custom default filesystem.

Not in scope for this PR though.

eirbjo commented:
> This should probably be mentioned in the compatibility risk section of the CSR, and in the release note too.

I have updated both to mention code relying on behavior of a failing `FtpURLConnection`.

eirbjo commented:
> > java/net/URLConnection/UNCTest.java

> This test wants to verify that any URLConnection to an UNC path is unconnected. Instead, since the UNC path does not exists, it just tests that the FtpURLConnection is unconnected, which it always is and has been and which in any case seems unrelated to JDK-4401485

Files https://bugs.openjdk.org/browse/JDK-8355475 to track an enhancement to make UNCTest use an existing Administrative Shares UNC path.

eirbjo commented:
> You will still need to update `java/net/URL/OpenStream.java` to make it pass with the new changes. Possibly have two `@run`, one with `-Djdk.net.file.ftpfallback=true` and one without and adapt the test expectation based on the presence (and value) of the property.

Hmm yes..

However, I find it somewhat dubious that 8202708 is crammed into this test which originally was written to test the non-proxy aspect in 4064962. Using that test to also verify that a non-existing `file://` UNC path leads to an FtpURLConnection which then fails with an `UnknownHostException` seems a bit convoluted and a mixing of concerns.

I think I would feel better if this aspect was instead verified in a new test added to the recently introduced `NonLocalFtpFallback`:

```java
/**
 * Verify that opening a stream on a non-proxy URLConnection for a non-local
 * file URL which is not an existing UNC path fails with UnknownHostException
 * when the fallback FtpURLConnection attempts to connect to the non-existing
 * FTP server.
 *
 * See bug 8202708
 */
public void verifyFtpConnectionException() throws IOException {
    URL url  = new URL("file://h7qbp368oix47/not-exist.txt");
    assertThrows(UnknownHostException.class, () -> {
        InputStream in = url.openConnection(Proxy.NO_PROXY).getInputStream();
    });
}
```

We could then remove any FTP aspect from `OpenStream.java` by reverting 8202708 and keep any FTP fallback testing in the recently introduced tests (disable and not disabled).

eirbjo commented:
>> You will still need to update java/net/URL/OpenStream.java to make it pass with the new changes

> I think I would feel better if this aspect was instead verified in a new test added to the recently introduced `NonLocalFtpFallback`

@dfuch I went ahead and implemented this in 5b11c07, could you have a look?

Should be easy to revert if you are opposed to this solution for fixing the currently failing `OpenStream` test.

dfuch commented:
> @dfuch I went ahead and implemented this in [5b11c07](https://github.com/openjdk/jdk/commit/5b11c073190a2bc42adf2a789407bbea9e100e7c), could you have a look?
> 
> Should be easy to revert if you are opposed to this solution for fixing the currently failing `OpenStream` test.

LGTM

eirbjo commented:
> I wonder if we need to add a comment in OpenStream that "bad" UNC paths are tested in NonLocalFtpFallback.

We could. But there is technincally such thing as a "bad" UNC path though. If File::exists fails, then it isn't considered a UNC path, so would fall back to FtpURLConnection. But so would the Unix implementation of `Handler::openConnection`.

So this testing is verifying network exceptions happing from FtpURLConnection. It isn't really UNC nor Windows-specific. With this in mind, the existing `OpenStream::testUnc` seems somewhat suspicious, so I'm worried a comment would just be confusing, at least if mentioning UNC.

eirbjo commented:
>  I wonder if we need to add a comment in OpenStream that "bad" UNC paths are tested in NonLocalFtpFallback.

Could we say something like: _For testing of non-local file URLs with the legacy FTP fallback feature enabled, see NonLocalFtpFallback._

dfuch commented:
> @dfuch Would you consider giving these updated tests a spin on Oracle CI to verify?

Will do. The CSR could also be moved to "Proposed".

eirbjo commented:
> Will do.

Great!

> The CSR could also be moved to "Proposed".

It was earlier today.

dfuch commented:
> I have added a Release Note as a subtask in the JBS issue, this also needs a review.

Hi Eirik, I have updated the wording of the release note at https://bugs.openjdk.org/browse/JDK-8354658 ; @AlanBateman or @Michael-Mc-Mahon may need to review my changes.

dfuch commented:
> @dfuch Would you consider giving these updated tests a spin on Oracle CI to verify?

The results came back green.

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java:
```diff
@@ -46,6 +47,12 @@ public class FileURLConnection extends URLConnection {
     private static final String TEXT_PLAIN = "text/plain";
     private static final String LAST_MODIFIED = "last-modified";
 
+    // The feature of falling back to FTP for non-local file URLs is disabled
+    // by default and can be re-enabled by setting a system property
+    private static boolean FTP_FALLBACK_ENABLED = "true".equalsIgnoreCase(
+            System.getProperty("jdk.net.file.ftpfallback", "false")
+    );
+
```
Maybe we should use `Boolean.getBoolean("jdk.net.file.ftpfallback")` here.

eirbjo replied:
I think I did it this way because I wanted the default to be explicit. But, yes Boolean::getBoolean is short and sweet.

dfuch commented on test/jdk/sun/net/www/protocol/file/NonLocalFtpFallbackDisabled.java:
```diff
@@ -0,0 +1,64 @@
+        URL localURL = file.toUri().toURL();
+        // We can use a fake host name here, no actual FTP request will be made
+        String hostname = "remotehost";
+        URL nonLocalURL = new URL("file", hostname, localURL.getFile());
```
Should we also check that `new URL("file", hostname, "").openConnection()` will throw MUE?

eirbjo replied:
Not sure why, but can't think of a reason why not. So added that plus avoided the deprectated URL constructor.

dfuch reviewed (commented):
Sorry - rewording the release note got me thinking again...

dfuch commented on test/jdk/sun/net/www/protocol/file/NonLocalFtpFallbackDisabled.java:
```diff
@@ -0,0 +1,71 @@
+
+        assertThrows(MalformedURLException.class, () -> {
+            // Non-empty path
+            URL localURL = file.toUri().toURL();
+            URL nonLocalURL = new URI("file", hostname, localURL.getFile(), "").toURL();
```
Actually we should take that out of the assertThrows since MalformedURLException could be thrown while constructing the URL...

eirbjo replied:
> MalformedURLException could be thrown while constructing the URL...

Good catch, fixed.

eirbjo commented:
/integrate delegate

Michael-Mc-Mahon commented:
> > I have added a Release Note as a subtask in the JBS issue, this also needs a review.
> 
> Hi Eirik, I have updated the wording of the release note at https://bugs.openjdk.org/browse/JDK-8354658 ; @AlanBateman or @Michael-Mc-Mahon may need to review my changes.

I fixed a few typos. Otherwise, it looks good.

eirbjo commented:
> > > I have added a Release Note as a subtask in the JBS issue, this also needs a review.
> > 
> > 
> > Hi Eirik, I have updated the wording of the release note at https://bugs.openjdk.org/browse/JDK-8354658 ; @AlanBateman or @Michael-Mc-Mahon may need to review my changes.
> 
> I fixed a few typos. Otherwise, it looks good.

Thanks Michael for your updates to the release note.

The property specification includes a recommendation that the ftp:// scheme should be used directly instead of relying on the file: scheme for FTP support.


Should we add some wording to that effect in the release note as well, to notify users that code relying on this legacy feature to resolve files using FTP should use explicit ftp:// URLs instead?

Note that I’m away from keyboard so cannot update JBS myself.

wenshao commented on src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java:
```diff
@@ -46,6 +47,11 @@ public class FileURLConnection extends URLConnection {
 
+    // The feature of falling back to FTP for non-local file URLs is disabled
+    // by default and can be re-enabled by setting a system property
+    private static boolean FTP_FALLBACK_ENABLED =
+            Boolean.getBoolean("jdk.net.file.ftpfallback");
```
```suggestion
    private static final boolean FTP_FALLBACK_ENABLED =
            Boolean.getBoolean("jdk.net.file.ftpfallback");
```
Add final here

eirbjo commented:
The CSR for this change has been approved. The Release Note has also been updated by more than one reviewer.

I don’t have any outstanding work for this PR on my side, unless there is more input from reviewers.

eirbjo commented:
Thanks for your patient help Daniel!

I’ll let this behavioral change in an ancient and dusty corner of the JDK linger for another 24h, just out of caution for any late reviewers.

eirbjo commented:
/integrate
